### ISSUE TITLE
How does the image reconstruction work?

### ISSUE THREAD

**Issue by dunky11** [zsyzzsoft/co-mod-gan#9]
First of all congrats on your results, your network is state of the art for image impainting by a margin...
Im currently trying to implement this in PyTorch and I am stuck because my TensorFlow 1 knowledge is limited.

As far as I understood it, your implementation is based on StyleGAN 2, however instead of just using a Generator with a mapping network to calculate the style vectors, you use a mapping network and a encoder network to calculate the style  vectors, where the style vector is calculated by applying a linear layer onto the concatenation of the output of  the encoder - E(y) and the Mapping network M(z). Here y is the image with holes which we want to reconstruct and z is noise. The architecture of the Encoder  is really similar to that of the discriminator in StyleGAN 2.

However does this really suffice to reconstruct the image? Each image in the batch gets mapped by the encoder into a 512/1024 dimensional vector, and this is the only information the network gets in order to impaint the image. Im feeling like I am missing a key piece here...

Thanks for any help in advance!

**Comment by zsyzzsoft** [comment]
There are also skip connections between the encoder and the decoder (like a U-Net) that provide multi-level spatial information, apart from the modulations that control global styles.

**Comment by dunky11** [comment]
So there are basically 4 networks: A discriminator, an encoder which is similar to the discriminator, which maps the masked image to a 512/1024 dimensional vector, a mapping network which maps noise to a 512/1024 dimensional vector and the generator, which takes the masked image as input and the output of the encoder/mapping network at each convolution.
The Generator is different from StyleGAN-2 because it is structured like a UNet, is that right?

**Comment by zsyzzsoft** [comment]
The generator consists of an encoder and a decoder (like a U-Net), and the encoder not only passes spatial information to the decoder but also maps the masked image to a style vector.

**Comment by dunky11** [comment]
So the encoder maps for example a 3x512x512 image to a 512x4x4 representation and a style vector.
The decoder then takes that 512x4x4 image and maps that back to 3x512x512 using the style vector and output of the mapping network at each convolution.

**Comment by dunky11** [comment]
I thinks i understood it now, thank you very much!

**Comment by Johnson-yue** [comment]
@dunky11  did you successfull implement code in PyTorch version?
